### DISCUSSION TITLE
Automatic Sleep Timer sometimes does not enable

### DISCUSSION THREAD

**Discussion by chrishoage** [advplyr/audiobookshelf-app#636] [Q&A]
On Android (Samsung Tab S5e) I have the automatic sleep timer enabled and set to 10 minutes. 

As the audio fades out I pause and pause to reset the timer. 

_Mostly_ the timer resets. 

_Sometimes_ the timer does not reset, instead disabled and the sleep timer does not come back on until I manually start the timer. 

Is there some behavior with the automatic sleep timer I am not understanding? 

Is the a setting I need to tweak? 

Basically the behavior I am after in the automatic sleep timer is: 

When in the auto sleep timer window always enable the sleep timer. 

If paused and resumed, enable the sleep timer.
If playing a new session, enable the sleep timer. 
If the sleep timer is disabled, paused, resumed, enable the sleep timer. 

I want to not have to check to make sure that the sleep timer is actually enabled. 

Is there a way to do this or is there a bug I am encountering? 

I don't have consistent steps to reproduce this yet. I just noted that sometimes I wake up (much longer after 10 minutes) and the book is still playing.

**Comment by advplyr** [comment]
This may be a bug because what you described is exactly how it should work now

**Reply by chrishoage** [reply]
I'll try to keep an eye out for specific steps to reproduce. 

As a developer I certainly understand how unhelpful "sometimes it doesn't work" reports are :sweat_smile:

**Reply by advplyr** [reply]
Hah, no problem. I know you were clarifying the expected behavior first.

**Reply by advplyr** [reply]
One thing that is producing unexpected results on some devices is battery optimization. You may want to try disabling battery optimization for the app.

**Reply by chrishoage** [reply]
I'll give that a shot this evening. 

One thing that kept happening last night is no matter what I did, the auto sleep timer wouldn't enable - at least visually - I didn't trust it enough to see if it was just a visual bug. 

Once I manually enabled it in the player it would re-enable during the pause and play. 

Still working on tracking down steps to reproduce. However my experience has been inconsistent so it's been challenging.

**Reply by chrishoage** [reply]
Turning off battery optimization did not seem to help. 

I'm having trouble getting exact steps to reproduce, but it seems like it's related to pause/resume or just resume from the lock screen and not the app. Anecdotally it seems that if I unlock and use the play button it seems to happen less than when using the lock screen. 

@advplyr I see logs in `SleepTimerManager.kt`. Is there a way to access these logs from the app? I don't see anything in the settings but perhaps I missed it. It might be helpful if I can examine these to see if anything seems wrong when I can reproduce this.

**Reply by chrishoage** [reply]
I could be misinterpreting how `Calendar.getInstance()` works (I'm not an Android dev) however I'm curious if I'm reading this right. 

- Start Calendar is set to right now 
- Start calendar is adjusted to the stored start hour and minute of the day the instance was created 
- End Calendar is set to right now 
- End calendar is adjusted to the stored end hour and minute of the day the instance was created 
- Check that if the end datetime is before the start datetime add a day to the end calendar

It seems like this logic might not be right. 

- Assume Start is set to 22:00
- Assume End is set to 8:00
- Assume `checkAutoSleepTimer` is run at 2023-03-27-00:30
- Start calendar is 2023-03-27-22:00
- End Calendar is 2023-03-27-08:00
- 2023-03-27-08:00 is before  2023-03-27-22:00
- End calendar becomes 2023-03-28-08:00
- Window is now Start: 2023-03-27-22:00 - End: 2023-03-28-08:00
- Current time is 2023-03-27-00:30
- Current time is outside of start window
- auto sleep time is not started

With out logs I can't verify this, but I believe this is a logic bug and is causing my sleep timer to not reset after midnight. 



```kt
      val startCalendar = Calendar.getInstance()
      startCalendar.set(Calendar.HOUR_OF_DAY, deviceSettings.autoSleepTimerStartHour)
      startCalendar.set(Calendar.MINUTE, deviceSettings.autoSleepTimerStartMinute)
      val endCalendar = Calendar.getInstance()
      endCalendar.set(Calendar.HOUR_OF_DAY, deviceSettings.autoSleepTimerEndHour)
      endCalendar.set(Calendar.MINUTE, deviceSettings.autoSleepTimerEndMinute)

      // In cases where end time is earlier then start time then we add a day to end time
      //   e.g. start time 22:00 and end time 06:00. End time will be treated as 6am the next day.
      //   e.g. start time 08:00 and end time 22:00. Start and end time will be the same day.
      if (endCalendar.before(startCalendar)) {
        endCalendar.add(Calendar.DAY_OF_MONTH, 1)
      }
```

**Reply by advplyr** [reply]
I'm not sure what you mean by the logic not being right. I was following you until 
```
Current time is outside of start window
```
In your example the current time is AFTER the start time and BEFORE the end time, which would be inside the window.
The bug may be here still I'm just not sure where you saw the logic flaw.

**Reply by advplyr** [reply]
I just re-read it and I see what you mean now. With `2023-03-27 00:30` this would be outside the window, you're right. Let me see if I can reproduce this.

**Reply by chrishoage** [reply]
Looks like you might be on this - just in case though I was able to reproduce this behavior on my phone (not near my tablet at the moment but I assume that distinction is irrelevant). 

Steps to Repo:

1. Consider current time (in this example it will be 2:20PM)
2. Ensure End Time is after Current Time (3:00PM)
3. Ensure Start Time is after End Time (10:00PM)
4. Play 
5. Observe timer state (not enabled)
6. Pause
7. Change Start Time to be before Current time (2:00 PM)
8. Change End time to be before Start Time (8:00 AM)
9. Play
10. Observe timer state (enabled)

**Reply by advplyr** [reply]
Thanks for tracking this down. I may be overthinking this but this solution worked in my testing.

```js
      // In cases where end time is before start time then we shift the time window forward or backward based on the current time.
      //   e.g. start time 22:00 and end time 06:00.
      //          If current time is less than start time (e.g. 00:30) then start time will be the previous day.
      //          If current time is greater than start time (e.g. 23:00) then end time will be the next day.
      if (endCalendar.before(startCalendar)) {
        if (currentCalendar.before(startCalendar)) { // Shift start back a day
          startCalendar.add(Calendar.DAY_OF_MONTH, -1)
        } else { // Shift end forward a day
          endCalendar.add(Calendar.DAY_OF_MONTH, 1)
        }
      }
```

https://github.com/advplyr/audiobookshelf-app/issues/648

**Reply by chrishoage** [reply]
> Thanks for tracking this down

Thank you for fixing it! 

> I may be overthinking this but this solution worked in my testing.

I think the logic might be able to be a little simpler (if you desire) 

- **let** current datetime be now
- **let** end datetime be current datetime-08:00
- **let** start datetime be current datetime-22:00
- **if** current datetime is before end datetime **set** start datetime as current date - 1 
- **else** **set** end datetime as current date + 1

Proof:

Current time is before end time and after start time

- **let** current datetime be 03-27-00:30
- **let** start datetime be 03-27-22:00
- **let** end datetime be 03-27-08:00
- start: 03-26-22:00 end: 03-27-08:00 current: 03-27-00:30 **timer on**

Current time is after end time and after start time

- **let** current datetime be 03-27-10:30
- **let** start datetime be 03-27-22:00
- **let** end datetime be 03-27-08:00
- start: 03-27-22:00 end: 03-28-08:00 current: 03-27-10:30 **timer off**

Current time is after start time and before end time

- **let** current datetime be 03-27-23:30
- **let** start datetime be 03-27-22:00
- **let** end datetime be 03-27-08:00
- start: 03-27-22:00 end: 03-28-08:00 current: 03-27-23:30 **timer on**

Start and End time are the same value

- **let** current datetime be 03-27-10:30
- **let** start datetime be 03-27-08:00
- **let** end datetime be 03-27-08:00
- start: 03-27-08:00 end: 03-28-08:00 current: 03-27-10:30 **timer on**

Current Time and Start Time are the same value

- **let** current datetime be 03-27-22:00
- **let** start datetime be 03-27-22:00
- **let** end datetime be 03-27-08:00
- start: 03-27-22:00 end: 03-28-08:00 current: 03-27-22:00 **timer on**

Current Time and End Time are the same value

- **let** current datetime be 03-27-08:00
- **let** start datetime be 03-27-22:00
- **let** end datetime be 03-27-08:00
- start: 03-28-22:00 end: 03-27-08:00 current:  03-27-08:00 **timer off**


So with this I _think_ the logic can just be 

```kt
      // In cases where current time is before end time then we shift the time window forward or backward based on the current time.
      //   e.g. start time 22:00 and end time 06:00.
      //          If current time is less than start time (e.g. 00:30) then start time will be the previous day.
      //          If current time is greater than start time (e.g. 23:00) then end time will be the next day.
      if (currentCalendar.before(endCalendar)) { // Shift start back a day
        startCalendar.add(Calendar.DAY_OF_MONTH, -1)
      } else { // Shift end forward a day
        endCalendar.add(Calendar.DAY_OF_MONTH, 1)
      }
```

This is of course entirely untested, so use your best judgement :) 

(as an aside this seems like a really good area to unit test one day)

**Reply by advplyr** [reply]
- let current datetime be 03-27-00:30
- let start datetime be 03-27-08:00
- let end datetime be 03-27-22:00
- start: 03-26-08:00 end: 03-27-22:00 current: 03-27-00:30 **timer on**

I think this doesn't work when the start time is before the end time

**Reply by chrishoage** [reply]
Whoops you're right! I totally overlooked that case.
